raiden_contracts/contracts/MonitoringService.sol

```diff
@@ -0,0 +1,321 @@
+        // Only allowed to claim, if channel is settled
+        // Channel is settled if it's data has been deleted
+        uint256 channel_settle_block;
+        (, channel_settle_block, ) = token_network.getChannelInfo(closing_participant, non_closing_participant);
+        require(channel_settle_block == 0);
+
+        Reward storage reward = rewards[channel_identifier];
+
+        // Make sure that the Reward exists
+        require(reward.reward_sender_address != 0x0);
+
+        address raiden_node_address = recoverAddressFromRewardProof(
+            channel_identifier,
+            reward_amount,
+            token_network_address,
+            token_network.chain_id(),
+            reward.nonce,
+            monitor_address,
+            reward.reward_proof_signature
+        );
+
+        // Deduct reward from raiden_node deposit
+        balances[raiden_node_address] -= reward_amount;
+        // Add reward to the monitoring services' balance.
+        // This minimizes the amount of gas cost
+        // Only use token.transfer in the withdraw function
+        balances[monitor_address] += reward_amount;
+
+        emit RewardClaimed(monitor_address, reward_amount, channel_identifier);
+
+        // delete storage
+        delete rewards[channel_identifier];
+    }
+
+    /// @notice Withdraw depositted tokens. If msg.sender is MS don't allow to
+    /// withdraw below minimum_deposit.
+    /// Can be called by addresses with a deposit as long as they have a positive balance
+    /// @param amount Amount of tokens to be withdrawn
+    function withdraw(uint amount) public {
+        // TODO: make sure that deposits reserved for open rewards cannot be withdrawn
+        if (registered_monitoring_services[msg.sender]) {
+            require(balances[msg.sender] - amount >= minimum_deposit);
+            balances[msg.sender] -= amount;
+            require(token.transfer(msg.sender, amount));
+        } else {
+            require(balances[msg.sender] - amount >= 0);
+            balances[msg.sender] -= amount;
+            require(token.transfer(msg.sender, amount));
+        }
+        emit Withdrawn(msg.sender, amount);
+    }
+
+    /// @notice Allow an address with a balance equal to or above minimum_deposit to
+    /// register as Monitoring Service.
+    function registerMonitoringService() isNotMonitor(msg.sender) public returns (bool) {
+        // only allow to register if not already registered
+        require(!registered_monitoring_services[msg.sender]);
+        if (balances[msg.sender] >= minimum_deposit) {
+            registered_monitoring_services[msg.sender] = true;
+            emit MonitoringServiceRegistered(msg.sender);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /// @notice Let Monitoring Services deregister as monitoring services.
+    /// Only callable by registered Monitoring Services
+    function deregisterMonitoringService() isMonitor(msg.sender) public {
```

Comment:
I would either make this function also transfer all the tokens to the MS, or include this logic into the `withdraw` for MSs (if MS wants to withdraw all tokens).